### PR TITLE
4 add sampling functions

### LINKED ISSUE #4: subset EPC data by LSOA
Select sample from EPC data: subset by LSOA with 2 LSOAs per devolved nation, one with min and one with max rows per LSOA.

### PR DESCRIPTION
**Pull request by crispy-wonton**
Fixes #4 

# Description
Changes:
- add `sampling.py` with sampling functions to subset EPC dataset by LSOA. 2 LSOAs are sampled per nation: one with max UPRN count and one with min UPRN count.
- add new feature processing functions and move existing ones from `getters.get_datasets.py` to `pipeline.prepare_features.py`
- rename `enhance_epc_functions.py` to `enhance_epc.py`

# Instructions for Reviewer

In order to test the code in this PR you need to run the following script in a notebook:
```
import polars as pl
from asf_heat_pump_suitability.pipeline import run_script, sampling

epc_path = "s3://asf-daps/lakehouse/processed/epc/deduplicated/processed_dedupl-0.parquet"

test = run_script.main(epc_path=epc_path)
sample = sampling.filter_df_area_sample_epc(test)
```

Please could you:
- check for errors/bugs issues
- check output works and is sensible. Output should include data from 6 LSOAs: 2 for each nation with min and max UPRN counts.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: asf_heat_pump_suitability/pipeline/prepare_features.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    Args
+        df (pl.DataFrame): dataset with LSOA columns
+
+    Returns
+        pl.DataFrame: dataset with new `lsoa` column
+
+    """
+    df = df.with_columns(pl.col(["lsoa11", "lsoa21"]).replace("", None))
+    df = df.with_columns(pl.col("lsoa21").fill_null(pl.col("lsoa11")).alias("lsoa"))
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/prepare_features.py]: Might we want the same done with MSOA?

File: asf_heat_pump_suitability/pipeline/sampling.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        df (pl.DataFrame): dataset to sample from
+        country_codes (list): list of country codes to sample from
+
+    Returns
+        list[dict]: list of dicts containing sample information
+    """
+    return [
+        df.filter(pl.col("country_code") == cc).sample(1, seed=8).to_dicts()[0]
+        for cc in country_codes
+    ]
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/sampling.py:73]: I would turn this into a script that can be run with

```
if __name__ == '__main__':
    epc_path = "s3://asf-daps/lakehouse/processed/epc/deduplicated/processed_dedupl-0.parquet"

    test = run_script.main(epc_path=epc_path, save_output=None)
    sample = sampling.filter_df_area_sample_epc(test)
    
    # Save sample on S3 somewhere
    ...
```

Comment (lizgzil) [reply asf_heat_pump_suitability/pipeline/sampling.py:73]: I guess you'll need `from asf_heat_pump_suitability.pipeline import run_script` too

**Review by lizgzil — COMMENTED** [review]
This looks great and the command given ran and produced a sample that I think looks as it should! I had to turn it into a pandas df to check it though since I don't know any of the polars commands 😆 

I guess in terms of turning this repo into a clearly documented and reproducible pipeline, I would add the following.
- turn the sampling.py into a script to be run (rather than running the commands in a notebook)
- save the sample output (seems unnecessary to have to resample every time when we could just be reading this dataset)
- add to a pipeline README e.g. "To create and save a sample run `python asf_heat_pump_suitability/pipeline/sampling.py`" (I don't really think it's necessary at this stage to spend time adding arguments since this sampling is such an ad hoc part of the full pipeline)

Another point was about MSOA. Since one of the datasets you found was per MSOA - should we perhaps save a sample per MSOA too? I guess you've made this so easy to create with `sampling.filter_df_area_sample_epc(test, area='msoa')` which is great!

File: asf_heat_pump_suitability/pipeline/prepare_features.py (hunk 1 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/prepare_features.py]: updated function with new `area_type` param to allow inputting different output area types.

File: asf_heat_pump_suitability/pipeline/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# asf_heat_pump_suitability.pipeline
```
Comment (lizgzil) [inline asf_heat_pump_suitability/pipeline/README.md:1]: lovely :)

**Review by lizgzil — DISMISSED** [review]
this looks great and then script ran. Thanks for making the changes :)

File: asf_heat_pump_suitability/pipeline/README.md (hunk 3 above)
Comment (crispy-wonton) [reply asf_heat_pump_suitability/pipeline/README.md:1]: @lizgzil I've added a new section to this README with instructions for running `run_script.py`

**Pull request merged**